enumerable.rb

```diff
@@ -1,87 +1,163 @@
+# frozen_string_literal: true
+
+# rubocop: disable Metrics/ModuleLength
+# rubocop: disable Metrics/PerceivedComplexity, Metrics/CyclomaticComplexity
+
+# module Enumerable
 module Enumerable
   def my_each
-    for i in self do
-        self.class == Array ? yield(i) : yield(i[0], i[1])
+    return to_enum(:my_each) unless block_given?
+
+    i = 0
+    while i < to_a.length
+      instance_of?(Array) ? yield(to_a[i]) : yield(to_a[i][0], to_a[i][1])
+      i += 1
     end
+    self
   end
 
   def my_each_with_index
-    for i in self do
-      self.class == Array ? yield(i, self.index(i)) : yield([i[0], i[1]], self.keys.index(i[0]))
+    return to_enum(:my_each_with_index) unless block_given?
+
+    i = 0
+    while i < to_a.length
+      instance_of?(Array) ? yield(to_a[i], i) : yield([to_a[i][0], to_a[i][1]], i)
+      i += 1
     end
+    self
   end
 
   def my_select
+    return to_enum(:my_select) unless block_given?
+
     new_arr = []
     new_hash = {}
-    self.my_each do |k,v|
-        if self.class == Array
-            new_arr << k if yield(k) == true
-        else
-            new_hash.store(k,v) if yield(k,v) == true
-        end
+    my_each do |k, v|
+      if instance_of?(Array)
+        new_arr << k if yield(k) == true
+      elsif yield(k, v) == true
+        new_hash.store(k, v)
+      end
     end
     new_arr.empty? ? new_hash : new_arr
   end
```

Comment:
- [ ]returns an array containing all elements of enum for which the given block returns a true value::range
```
  range = Range.new(1,10)
  block =  proc { |num| num < (0 + 9) / 2 }
  p range.my_select(&block)  # gotten {1=>0, 0=>1}

  but
  
  range = Range.new(1,10)
  block =  proc { |num| num < (0 + 9) / 2 }
  p range.select(&block)  # is [1, 2, 3]
```